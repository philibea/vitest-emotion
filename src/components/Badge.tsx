import type { Theme } from '@emotion/react';
import { useTheme } from '@emotion/react';
import styled from '@emotion/styled';
import type { ReactNode } from 'react';
import { useMemo } from 'react';

const TEXT_SIZES = {
  large: 14,
  medium: 12,
  small: 10,
};

export const SIZES = {
  large: 32,
  medium: 24,
  small: 16,
};

export const PROMINENCES = {
  default: 'default',
  strong: 'strong',
};

/**
 * Generate all styles available for badge based on prominence and sentiments
 */
const generateStyles = ({
  prominence,
  theme,
}: {
  prominence: string;
  theme: Theme;
}): Record<'disabled' | 'neutral', string> => ({
  disabled: `
      color: ${theme.colors.neutral.textWeak};
      background: ${theme.colors.neutral.backgroundStrong};
      border: none;
    `,
  neutral: `
      color: ${
        prominence === PROMINENCES.strong
          ? theme.colors.neutral.textStronger
          : theme.colors.neutral.text
      };
      background: ${
        prominence === PROMINENCES.strong
          ? theme.colors.neutral.backgroundStronger
          : theme.colors.neutral.backgroundWeak
      };
      border: 1px solid ${
        prominence === PROMINENCES.strong
          ? theme.colors.neutral.borderStronger
          : theme.colors.neutral.border
      };
    `,
});

const StyledSpan = styled.div<{
  size: number;
  sentimentStyles: string;
  fontSize: number;
}>`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  white-space: nowrap;
  border-radius: ${({ theme }) => theme.radii.xlarge};
  padding: 0
    ${({ theme, size }) =>
      size === SIZES.small ? theme.space['1'] : theme.space['2']};
  gap: ${({ theme, size }) =>
    size === SIZES.small ? theme.space['0.5'] : theme.space['1']};
  width: fit-content;
  height: ${({ size }) => size}px;
  text-transform: uppercase;
  font-size: ${({ fontSize }) => fontSize}px;
  color: inherit;
  ${({ sentimentStyles }) => sentimentStyles}
`;

type BadgeProps = {
  size?: keyof typeof SIZES;
  prominence?: keyof typeof PROMINENCES;
  /**
   * Defines icon to display on left side of badge. **Only available on medium and large sizes**.
   */
  disabled?: boolean;
  className?: string;
  children: ReactNode;
  'data-testid'?: string;
};

/**
 * Badge component is used to display a status or a label in a small container.
 */
export const Badge = ({
  // sentiment = 'neutral',
  size = 'medium',
  prominence = 'default',
  // disabled = false,
  className,
  children,
  'data-testid': dataTestId,
}: BadgeProps) => {
  const theme = useTheme();

  const generatedStyles = useMemo(
    () => generateStyles({ prominence, theme }),
    [prominence, theme],
  );

  const sizeValue = SIZES[size];

  return (
    <StyledSpan
      // aria-label={ariaLabel}
      sentimentStyles={generatedStyles['disabled']}
      size={sizeValue}
      // variant="bodyStrong"
      as="span"
      fontSize={TEXT_SIZES[size]}
      // prominence={disabled ? 'weak' : 'default'}
      className={className}
      data-testid={dataTestId}
    >
      {children}
    </StyledSpan>
  );
};
