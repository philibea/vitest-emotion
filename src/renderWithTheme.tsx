import React from 'react';
import createCache from '@emotion/cache';
import { createSerializer } from '@emotion/jest';
import { render } from '@testing-library/react';
import { consoleLightTheme } from '@ultraviolet/themes';
import { CacheProvider, ThemeProvider } from '@emotion/react';

const emotionCache = createCache({
  key: 'cache',
});

emotionCache.compat = true;

export const renderWithTheme = (children: React.ReactNode) => {
  expect.addSnapshotSerializer(createSerializer());

  return render(
    <CacheProvider value={emotionCache}>
      <ThemeProvider theme={consoleLightTheme}>{children}</ThemeProvider>
    </CacheProvider>,
  );
};
