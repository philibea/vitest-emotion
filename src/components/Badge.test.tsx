import { it, describe, expect } from 'vitest';
import { renderWithTheme } from '../renderWithTheme';
import { Badge } from './Badge';

describe('Badge', () => {
  it('renders ', () => {
    const { container } = renderWithTheme(<Badge>Badge </Badge>);

    expect(container).toMatchSnapshot();
  });
  it('renders disabled ', () => {
    const { container } = renderWithTheme(<Badge disabled>Badge </Badge>);

    expect(container).toMatchSnapshot();
  });
});
