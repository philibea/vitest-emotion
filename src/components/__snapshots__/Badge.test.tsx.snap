// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Badge > renders  1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  white-space: nowrap;
  border-radius: 16px;
  padding: 0 16px;
  gap: 8px;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: 24px;
  text-transform: uppercase;
  font-size: 12px;
  color: inherit;
  color: #727683;
  background: #e9eaeb;
  border: none;
}

<div>
  <span
    class="emotion-0"
    font-size="12"
    size="24"
  >
    Badge 
  </span>
</div>
`;

exports[`Badge > renders disabled  1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  white-space: nowrap;
  border-radius: 16px;
  padding: 0 16px;
  gap: 8px;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: 24px;
  text-transform: uppercase;
  font-size: 12px;
  color: inherit;
  color: #727683;
  background: #e9eaeb;
  border: none;
}

.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  white-space: nowrap;
  border-radius: 16px;
  padding: 0 16px;
  gap: 8px;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: 24px;
  text-transform: uppercase;
  font-size: 12px;
  color: inherit;
  color: #727683;
  background: #e9eaeb;
  border: none;
}

<div>
  <span
    class="emotion-0"
    font-size="12"
    size="24"
  >
    Badge 
  </span>
</div>
`;
